### PR TITLE
CHelper_libXBMC_addon: Avoid buffer overflow in Log

### PR DESCRIPTION
**Pull request by palinek**
## Description

Simple change to avoid possible buffer overflow in log method.

## Motivation and Context

Do not allow plugin to crash kodi. E. g. kodi-pvr/pvr.iptvsimple#173

## How Has This Been Tested?

It builds and doesn't break API/ABI.

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by Jalle19 — APPROVED** [review]

**Comment by Jalle19** [comment]
Someone else should probably approve too if it's fine. @Paxxi ?

**Comment by a1rwulf** [comment]
I think it's good.
One could initialize the whole buffer to 0 then the `buffer[sizeof (buffer) - 1] = '\0';` wouldn't be necessary.
But it's already an approvement as is.

**Comment by palinek** [comment]
> One could initialize the whole buffer to 0 then the buffer[sizeof (buffer) - 1] = '\0'; wouldn't be necessary.

Does it make any difference in terms of speed?

**Comment by Paxxi** [comment]
Zeroing before is a waste of time because if the string fills the buffer you don't get a null terminator and bad things will happen.
Looks good to me

**Review by Paxxi — APPROVED** [review]

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h [hunk 1]
```diff
@@ -203,14 +203,17 @@ namespace ADDON
...
      */
     void Log(const addon_log_t loglevel, const char *format, ... )
     {
       char buffer[16384];
       va_list args;
       va_start (args, format);
-      vsprintf (buffer, format, args);
+      vsnprintf (buffer, sizeof (buffer), format, args);
       va_end (args);
+      buffer[sizeof (buffer) - 1] = '\0'; // to be sure it's nul-terminated
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: typo: `null-terminated`

Comment (palinek) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: I'm referring here to nul ASCII character, https://tools.ietf.org/html/rfc20#section-2, not to null pointer.

But if you wish, I can change it. Should I?

Comment (Rechi) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: Most (all?) documentation call it null termination, but you can keep it as it is.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h [hunk 2]
```diff
@@ -203,14 +203,17 @@ namespace ADDON
...
+     * @note This method uses limited buffer (16k) for the formatted output.
+     * So data, which will not fit into it, will be silently discarded.
      */
     void Log(const addon_log_t loglevel, const char *format, ... )
     {
       char buffer[16384];
       va_list args;
       va_start (args, format);
-      vsprintf (buffer, format, args);
+      vsnprintf (buffer, sizeof (buffer), format, args);
```
Comment (garbear) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: Is `vsnprintf()` like `strncpy()`? If so, this should be `vsnprintf(buffer, sizeof(buffer) - 1, format, args);`

Comment (palinek) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: right, but we can gain a 0.00001% speedup with the lower bound :-D. I'll change the commit to make everyone happy.

Comment (palinek) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/libXBMC_addon.h]: ... with the assumption that `sizeof(buffer) - 1` is evaluated in compile time

**Comment by Rechi** [comment]
jenkins build and merge

**Comment by a1rwulf** [comment]
thx a lot for your contribution @palinek

**Comment by palinek** [comment]
> thx a lot for your contribution

no problem

**Pull request merged**